cmd/kubelet/app/options/globalflags.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package options
+
+import (
+	"flag"
+	"os"
+	"strings"
+
+	"github.com/spf13/pflag"
+
+	// libs that provide registration functions
+	"k8s.io/apiserver/pkg/util/logs"
+	"k8s.io/kubernetes/pkg/version/verflag"
+
+	// ensure libs have a chance to globally register their flags
+	_ "github.com/golang/glog"
+	_ "github.com/google/cadvisor/container/common"
+	_ "github.com/google/cadvisor/container/containerd"
+	_ "github.com/google/cadvisor/container/docker"
+	_ "github.com/google/cadvisor/container/raw"
+	_ "github.com/google/cadvisor/machine"
+	_ "github.com/google/cadvisor/manager"
+	_ "github.com/google/cadvisor/storage"
+)
+
+// AddGlobalFlags explicitly registers flags that libraries (glog, verflag, etc.) register
+// against the global flagsets from "flag" and "github.com/spf13/pflag".
+// We do this in order to prevent unwanted flags from leaking into the Kubelet's flagset.
+func AddGlobalFlags(fs *pflag.FlagSet) {
+	addGlogFlags(fs)
+	addCadvisorFlags(fs)
+	verflag.AddFlags(fs)
+	logs.AddFlags(fs)
+}
+
+// register adds a flag to local that targets the Value associated with the Flag named globalName in global
+func register(global, local *flag.FlagSet, globalName, help string) {
+	if f := global.Lookup(globalName); f != nil {
+		// we should always use hyphens instead of underscores when registering kubelet flags
+		local.Var(f.Value, strings.Replace(globalName, "_", "-", -1), help)
+	}
+}
+
+// addGlogFlags adds flags from github.com/golang/glog
+func addGlogFlags(fs *pflag.FlagSet) {
+	// lookup flags in global flag set and re-register the values with our flagset
+	global := flag.CommandLine
+	local := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	register(global, local, "logtostderr", "log to standard error instead of files")
```

Comment:
when we get to the point of reparsing command lines, the fact that we're registering global values into local flagsets means that any accumulating Values will continue accumulating... it doesn't matter that we put them into a new flagset. we may have gotten lucky with all of these (though I wonder about ones like `vmodule`), but that's not a given

Reply:
That PR (#56995) extends register to allow creation of "fake" flagsets, where the values all have noop Set implementations so that we can avoid double-setting the global values: 
```
// register adds a flag to local that targets the Value associated with the Flag named globalName in global
// If fake is true, will register the flag name, but point it at a value with a noop Set implementation.
func register(global, local *flag.FlagSet, globalName, help string, fake bool) {
	if fake {
		local.Var(flagutil.NoOp{}, strings.Replace(globalName, "_", "-", -1), help)
	} else if f := global.Lookup(globalName); f != nil {
		// we should always use hyphens instead of underscores when registering kubelet flags
		local.Var(f.Value, strings.Replace(globalName, "_", "-", -1), help)
	}
}
```